topology/probes/ovs_of.go

```diff
@@ -0,0 +1,503 @@
+
+// OvsOfProbe is the type of the probe retrieving Openflow rules on an Open Vswitch
+type OvsOfProbe struct {
+	sync.Mutex
+	Host         string                    // The host
+	Graph        *graph.Graph              // The graph that will receive the rules found
+	Root         *graph.Node               // The root node of the host in the graph
+	BridgeProbes map[string]*BridgeOfProbe // The table of probes associated to each bridge
+	Translation  map[string]string         // A translation table to find the url for a given bridge knowing its name
+	Certificate  string                    // Path to the certificate used for authenticated communication with bridges
+	PrivateKey   string                    // Path of the private key authenticating the probe.
+	CA           string                    // Path of the certicate of the Certificate authority used for authenticated communication with bridges
+	sslOk        bool                      // cert private key and ca are provisionned.
+	Sudo         bool                      // Do we need sudo to launch the command on the bridge
+}
+
+// BridgeOfProbe is the type of the probe retrieving Openflow rules on a Bridge
+type BridgeOfProbe struct {
+	Host       string           // The global host
+	Bridge     string           // The bridge monitored
+	UUID       string           // The UUID of the bridge node
+	Address    string           // The address of the bridge if different from name
+	BridgeNode *graph.Node      // the bridge node on which the rule nodes are attached.
+	OvsOfProbe *OvsOfProbe      // Back pointer to the probe
+	Rules      map[string]*Rule // The set of rules found so far.
+	cancel     context.CancelFunc
+}
+
+// Rule is an OpenFlow rule in a switch
+type Rule struct {
+	Cookie  uint64 // cookie value of the rule
+	Table   int    // table containing the rule
+	Filter  string // all the filters as a comma separated string
+	Actions string // all the actions (comma separated)
+	InPort  int    // -1 is any
+	UUID    string // Unique id
+}
+
+// Event is an event as monitored by ovs-ofctl monitor <br> watch:
+type Event struct {
+	Rule   *Rule  // The rule modified
+	Date   int64  // the date of the event
+	Action string // the action taken
+	Bridge string // The bridge whtere it ocured
+}
+
+// ProtectCommas substitute commas with semicolon
+// when inside parenthesis
+func ProtectCommas(line string) string {
+	work := []rune(line)
+	var braces = 0
+	for i, word := range work {
+		switch word {
+		case '(':
+			braces++
+		case ')':
+			braces--
+		case ',':
+			if braces > 0 {
+				work[i] = ';'
+			}
+		}
+	}
+	return string(work)
+}
+
+// fillIn is a utility function that takes a splitted rule line
+// and fills a Rule/Event structure with it
+func fillIn(components []string, rule *Rule, event *Event) {
+	for _, component := range components {
+		keyvalue := strings.SplitN(component, "=", 2)
+		if len(keyvalue) == 2 {
+			key := keyvalue[0]
+			value := keyvalue[1]
+			switch key {
```

Comment:
Because unfortunately it is not provided when you listen to rule changes. You have to get the priority from another query on the installed rules.